cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMetadataStateManager.java

```diff
@@ -0,0 +1,385 @@
+
+    private final Reconciliation reconciliation;
+
+    private KafkaMetadataState metadataState;
+
+    private String kraftAnno;
+
+    private boolean isMigrationDone = false;
+
+    /**
+     * Constructor
+     *
+     * @param reconciliation Reconciliation information
+     * @param kafkaCr instance of the Kafka CR
+     * @param useKRaftFeatureGateEnabled if the UseKRaft feature gate is enabled on the operator
+     */
+    public KafkaMetadataStateManager(
+            Reconciliation reconciliation,
+            Kafka kafkaCr,
+            boolean useKRaftFeatureGateEnabled) {
+        this.reconciliation = reconciliation;
+        this.kraftAnno = kraftAnnotation(kafkaCr);
+        String metadataStateFromKafkaCr = kafkaCr.getStatus() != null ? kafkaCr.getStatus().getKafkaMetadataState() : null;
+        // missing metadata state means reconciling an already existing Kafka resource with newer operator supporting metadata state or first reconcile
+        if (metadataStateFromKafkaCr == null) {
+            this.metadataState = isKRaftAnnoEnabled() ? KafkaMetadataState.KRaft : KafkaMetadataState.ZooKeeper;
+        } else {
+            this.metadataState = KafkaMetadataState.valueOf(metadataStateFromKafkaCr);
+        }
+        if (!useKRaftFeatureGateEnabled && (isKRaftAnnoEnabled() || isKRaftAnnoMigration())) {
+            LOGGER.errorCr(reconciliation, "Trying to reconcile a KRaft enabled cluster or migrating to KRaft without the useKRaft feature gate enabled");
+            throw new IllegalArgumentException("Failed to reconcile a KRaft enabled cluster or migration to KRaft because useKRaft feature gate is disabled");
+        }
+        LOGGER.infoCr(reconciliation, "Loaded metadata state from the Kafka CR [{}] and strimzi.io/kraft [{}]", this.metadataState, this.kraftAnno);
+    }
+
+    /**
+     * Computes the next state in the metadata migration Finite State Machine (FSM)
+     * based on the current state and the strimzi.io/kraft annotation value at the
+     * beginning of the reconciliation when this instance is created
+     *
+     * @param kafkaStatus Status of the Kafka custom resource where warnings about any issues with metadata state will be added
+     *
+     * @return the next FSM state
+     */
+    public KafkaMetadataState computeNextMetadataState(KafkaStatus kafkaStatus) {
+        KafkaMetadataState currentState = metadataState;
+        metadataState = switch (currentState) {
+            case KRaft -> onKRaft(kafkaStatus);
+            case ZooKeeper -> onZooKeeper(kafkaStatus);
+            case KRaftMigration -> onKRaftMigration(kafkaStatus);
+            case KRaftDualWriting -> onKRaftDualWriting(kafkaStatus);
+            case KRaftPostMigration -> onKRaftPostMigration(kafkaStatus);
+            case PreKRaft -> onPreKRaft(kafkaStatus);
+        };
+        LOGGER.infoCr(reconciliation, "from [{}] to [{}] with strimzi.io/kraft: [{}]", currentState, metadataState, kraftAnno);
```

Comment:
We can add more text instead of just "from ... to ... ". Of course, I don't know it's staying in the same state, so it will be "from X to X ... " just showing no state change happened. Not sure we want to have a different logging message just for that, it would be pretty clear no transition happened. Any further suggestion to improve it?

Reply:
It is info level message. So I wonder if we should log it only when the state changes? Or move it to debug?

Reply:
While it's not important during normal execution so INFO can pollute the log, I think it's useful during migration. Not only when a transition happens but even where there is no transition but you know that the Kafka is a such a state, with that kraft annotation and something is going to happen. I can have two different log messages (i.e. "In state X ..." when no transition, "Transitioning from X to Y ..." when there is a transition) but also set DEBUG log level. In this case I would suggest people to change the log level to DEBUG when doing migration, maybe in the doc, wdyt?

Reply:
This let me think that even Kafka upstream community suggest to have TRACE level on migration related logging so that you get more info on controller nodes. Maybe this is another suggestion to make on the documentation? It's useful for debugging purposes.

Reply:
So something like ....

```java
if (metadataState != currentState) {
            LOGGER.debugCr(reconciliation, "Transitioning metadata state from [{}] to [{}] with strimzi.io/kraft annotation [{}]", currentState, metadataState, kraftAnno);
        } else {
            LOGGER.debugCr(reconciliation, "Metadata state [{}] with strimzi.io/kraft annotation [{}]", metadataState, kraftAnno);
        }
```

Reply:
I can see the info level being useful when the state is changing. But I do not see a need to log it all the time when it is not changing. So I would move that to debug. If you want to keep users aware of the ongoing migration, it might be useful to have some info or warning messages when it is ready to move to the next state.

So e.g.:
* Have an infor message `Transitioning metadata state from X to Y`
* Have a debug level message `Metadata state X with ...`
* When you want the user to make the next step, have a warning message saying something like `Ready to finish the migration -> please set strimzi.io/kraft annotation to enabled` or whatever is the corresponding step

Reply:
Yeah good idea ;-)